---
title: "BITFAM user guide"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# install dependent packages

## rstan 

rstan is the R interface to Stan. You can have more information at [rstan](https://mc-stan.org/rstan/) website.

### Installing from the repository

Ubuntu users on R>=3.5.0 can install a binary version of RStan with

```{bash, eval = FALSE}
# Add Michael Rutter's c2d4u3.5 PPA (and rrutter3.5 for CRAN builds too)
sudo add-apt-repository -y "ppa:marutter/rrutter3.5"
sudo add-apt-repository -y "ppa:marutter/c2d4u3.5"
sudo apt update
sudo apt install r-cran-rstan
```

### C++ toolchain configuration

The following will create or edit a configuration file for the C++ toolchain

```{r, eval = FALSE}
dotR <- file.path(Sys.getenv("HOME"), ".R")
if (!file.exists(dotR)) dir.create(dotR)
M <- file.path(dotR, "Makevars")
if (!file.exists(M)) file.create(M)
cat("\nCXX14FLAGS=-O3 -march=native -mtune=native -fPIC",
    "CXX14=g++", # or clang++ but you may need a version postfix
    file = M, sep = "\n", append = TRUE)
```

### Installing RStan from source

```{r, eval = FALSE}
install.packages("rstan", type = "source")
```

## Seurat 

Seurat is a toolkit for single cell data analysis. You can have more information at [Seurat] (https://satijalab.org/seurat/) website.

### installing Seurat

```{r, eval = FALSE}
install.packages('Seurat')
```

# Quick start

## installing BITFAM

```{r, eval = FALSE}
devtools::install_github("jalees/BITFAM")
```
```{r}
library(BITFAM)
```
## preprocessing data

This is an optional step. BITFAM required the normalized scRNA-seq data. If you have already normalized the data, you can go directly to the main function in the next step.

```{r, eval = FALSE}
data_matrix_normalized <- BITFAM_preprocess(raw_data = your_raw_data)
```

## infer the transcription factor activities

The function of BITFAM is to infer the transcription factor activities. BITFAM is available for mouse and human. Mouse and human will have different ChIP-seq target datasets. User should specify the species they are using. The names of genes should be offical gene symbols. For mouse, the gene names are first letter uppercase. For human, the gene names are all letter uppercases.

```{r}
head(rownames(data_matrix_normalized), n = 5)
```

By default, BITFAM will choose the transcription factors that are in the most variable expressed genes to learn. But the users are also allowed to add transcription factors that they are interested in. If the users have their own scATAC-seq data which could be a supplemental filters to select the protential target genes, they can input the preprocessed Seurat object of scATAC-seq data wiht the augment scATAC_obj. Please note that these two functions are still under developing and testing. We highly recommend the users to try the default setting and contact us if you meet any problems.  

Here we use the Tabula Muris lung data as an example to show the usage of BITFAM main function.

```{r}

BITFAM_res <- BITFAM(data = data_matrix_normalized, species = "mouse", interseted_TF = c("Pax5", "Tal1"), scATAC_obj = NA, ncores = 10)

```

Then we can get the inferred activities for every TF in each individual cell by function BITFAM_activities. The input is the result from BITFAM main function.

```{r}
Z <- BITFAM_activities(BITFAM_res)
```

A tsne plot on the inferred transcription factor activities. 

```{r}
library(ggplot2)
library(Rtsne)

Z_tsne <- as.data.frame(Rtsne(Z)$Y)
colnames(Z_tsne) <- c("tSNE_1", "tSNE_2")
Annotation <- read.csv(system.file("extdata", "Lung_droplet_annotation.csv", package = "BITFAM"), stringsAsFactors = F)
Z_tsne$Cell_type <- Annotation$cell_ontology_class
ggplot(Z_tsne, aes(tSNE_1, tSNE_2, color = Cell_type)) + geom_point(size=0.8)
```


